### PR TITLE
feat(evm-gateway-package): package for evm AmplifierGateway encoding

### PR DESCRIPTION
**Pull request by haiyizxx**
## Description
- automatically fetch external gateway inferface files if not exist
- the package generate evm gateway types in rust for ampd and multisig-prover to use 

## Todos

- [ ] Unit tests
- [ ] Manual tests
- [ ] Documentation
- [ ] Connect epics/issues

### CODE REVIEW

File: packages/evm-gateway/build.rs [hunk 1]
```diff
@@ -0,0 +1,78 @@
+use std::env;
+use std::fs::{self, File};
+use std::io;
+use std::path::{Path, PathBuf};
+
+use reqwest::blocking::Client;
+use zip::ZipArchive;
+
+const SOLIDITY_RELEASES_URL: &str =
```
Comment (fish-sammy) [inline packages/evm-gateway/build.rs:9]: This should also go into Cargo.toml

File: packages/evm-gateway/release.toml [hunk 2]
```diff
@@ -0,0 +1 @@
+release = false
```
Comment (fish-sammy) [inline packages/evm-gateway/release.toml:1]: What's this file?

File: packages/evm-gateway/build.rs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+    let mut zipfile = File::create(zip_path)?;
+    io::copy(&mut response, &mut zipfile)?;
+
+    let zipfile = File::open(zip_path)?;
+    Ok(ZipArchive::new(zipfile)?)
+}
+
+fn extract(archive: &mut ZipArchive<File>) -> io::Result<()> {
+    let abi_output = Path::new("src/abi");
```
Comment (fish-sammy) [inline packages/evm-gateway/build.rs:56]: Don't hard code the path.

File: packages/evm-gateway/build.rs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            "contracts/interfaces/{}.sol/{}",
+            abi.trim_end_matches(".json"),
+            abi
+        );
+        let output_path = abi_output.join(abi);
+
+        let mut file = archive.by_name(&file_path).map_err(|_| {
+            io::Error::new(
+                io::ErrorKind::NotFound,
+                format!("File not found in archive: {}", file_path),
```
Comment (fish-sammy) [inline packages/evm-gateway/build.rs:69]:
```suggestion
                format!("file not found in archive: {}", file_path),
```

File: packages/evm-gateway/build.rs [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    Ok(())
+}
+
+fn download(url: &str, zip_path: &Path) -> Result<ZipArchive<File>, Box<dyn std::error::Error>> {
+    let client = Client::new();
+    let mut response = client.get(url).send()?;
+    if !response.status().is_success() {
+        return Err(Box::new(io::Error::new(
+            io::ErrorKind::Other,
+            format!("Failed to download {}", url),
```
Comment (fish-sammy) [inline packages/evm-gateway/build.rs:44]:
```suggestion
            format!("failed to download {}", url),
```

File: packages/evm-gateway/build.rs [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+const ABI_FILES: [&str; 2] = ["IAxelarAmplifierGateway.json", "IBaseWeightedMultisig.json"];
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    if !cfg!(feature = "run-script") {
+        return Ok(());
+    }
+
+    let version_tag = env::var("SOLIDITY_GATEWAY_VERSION")
+        .expect("Environment variable SOLIDITY_GATEWAY_VERSION is not set");
```
Comment (fish-sammy) [inline packages/evm-gateway/build.rs:20]:
```suggestion
        .expect("environment variable SOLIDITY_GATEWAY_VERSION is not set");
```

**Review by fish-sammy — CHANGES_REQUESTED** [review]

File: packages/evm-gateway/release.toml (hunk 2 above)
Comment (haiyizxx) [reply packages/evm-gateway/release.toml:1]: used by release pipeline, disable releases for internal packages

File: .gitignore [hunk 7]
```diff
@@ -17,3 +17,6 @@
 *.iml
 .idea
 .vscode
+
+# EVM gateway artifacts
+packages/evm-gateway/src/abi/
```
Comment (fish-sammy) [inline .gitignore:22]: Create a `.gitignore` in the package itself with `src/abi`.

**Review by fish-sammy — APPROVED** [review]

**Pull request merged**